File: android/src/main/java/com/opentokreactnative/utils/EventUtils.java
package com.opentokreactnative.utils;

import com.facebook.react.bridge.Arguments;
import com.facebook.react.bridge.WritableMap;
import com.facebook.react.bridge.WritableArray;
import com.opentok.android.Connection;
import com.opentok.android.OpentokError;
import com.opentok.android.Session;
import com.opentok.android.Stream;
import com.opentok.android.SubscriberKit;
import com.opentok.android.PublisherKit;

public final class EventUtils {

    public static WritableMap prepareJSConnectionMap(Connection connection) {

        WritableMap connectionInfo = Arguments.createMap();
        if (connection != null) {
            connectionInfo.putString("connectionId", connection.getConnectionId());
            connectionInfo.putString("creationTime", connection.getCreationTime().toString());
            connectionInfo.putString("data", connection.getData());
        }
        return connectionInfo;
    }

    public static WritableMap prepareJSStreamMap(Stream stream, Session session) {

        WritableMap streamInfo = Arguments.createMap();
        if (stream != null) {
            streamInfo.putString("streamId", stream.getStreamId());
            streamInfo.putInt("height", stream.getVideoHeight());
            streamInfo.putInt("width", stream.getVideoWidth());
            streamInfo.putString("creationTime", stream.getCreationTime().toString());
            streamInfo.putString("connectionId", stream.getConnection().getConnectionId());
            streamInfo.putString("sessionId", session.getSessionId());
            streamInfo.putMap("connection", prepareJSConnectionMap(stream.getConnection()));
            streamInfo.putString("name", stream.getName());
            streamInfo.putBoolean("hasAudio", stream.hasAudio());
            streamInfo.putBoolean("hasVideo", stream.hasVideo());
            if (stream.getStreamVideoType().equals(Stream.StreamVideoType.StreamVideoTypeScreen)) {
                streamInfo.putString("videoType", "screen");
            } else {
                streamInfo.putString("videoType", "camera");
            }
        }
        return streamInfo;
    }

    public static WritableMap prepareJSErrorMap(OpentokError error) {

        WritableMap errorInfo = Arguments.createMap();
        errorInfo.putString("message", error.getMessage());
        errorInfo.putString("code", error.getErrorCode().toString());
        return errorInfo;
    }

    public static WritableMap prepareJSSessionMap(Session session) {

        WritableMap sessionInfo = Arguments.createMap();
        sessionInfo.putString("sessionId", session.getSessionId());
        if (session.getConnection() != null) {
            WritableMap connectionInfo = prepareJSConnectionMap(session.getConnection());
            sessionInfo.putMap("connection", connectionInfo);
        }
        return sessionInfo;
    }

    public static WritableMap prepareStreamPropertyChangedEventData(String changedProperty, String oldValue, String newValue, Stream stream, Session session) {

        WritableMap streamPropertyEventData = Arguments.createMap();
        streamPropertyEventData.putString("changedProperty", changedProperty);
        streamPropertyEventData.putString("oldValue", oldValue);
        streamPropertyEventData.putString("newValue", newValue);
        streamPropertyEventData.putMap("stream", prepareJSStreamMap(stream, session));
        return streamPropertyEventData;
    }

    public static WritableMap prepareStreamPropertyChangedEventData(String changedProperty, WritableMap oldValue, WritableMap newValue, Stream stream, Session session) {

        WritableMap streamPropertyEventData = Arguments.createMap();
        streamPropertyEventData.putString("changedProperty", changedProperty);
        streamPropertyEventData.putMap("oldValue", oldValue);
        streamPropertyEventData.putMap("newValue", newValue);
        streamPropertyEventData.putMap("stream", prepareJSStreamMap(stream, session));
        return streamPropertyEventData;
    }

    public static WritableMap prepareStreamPropertyChangedEventData(String changedProperty, Boolean oldValue, Boolean newValue, Stream stream, Session session) {

        WritableMap streamPropertyEventData = Arguments.createMap();
        streamPropertyEventData.putString("changedProperty", changedProperty);
        streamPropertyEventData.putBoolean("oldValue", oldValue);
        streamPropertyEventData.putBoolean("newValue", newValue);
        streamPropertyEventData.putMap("stream", prepareJSStreamMap(stream, session));
        return streamPropertyEventData;
    }

    public static WritableMap prepareSubscriberAudioNetworkStats(SubscriberKit.SubscriberAudioStats stats) {

        WritableMap audioStats = Arguments.createMap();
        audioStats.putInt("audioPacketsLost", stats.audioPacketsLost);
        audioStats.putInt("audioBytesReceived", stats.audioBytesReceived);
        audioStats.putInt("audioPacketsReceived", stats.audioPacketsReceived);
        return audioStats;
    }

    public static WritableMap preparePublisherAudioNetworkStats(PublisherKit.PublisherAudioStats[] stats) {

        WritableMap eventData = Arguments.createMap();
        WritableArray jsStatsList = Arguments.createArray();

        for (PublisherKit.PublisherAudioStats statsItem : stats) {
            WritableMap jsDataItem = Arguments.createMap();
            jsDataItem.putString("connectionId", statsItem.connectionId);
            jsDataItem.putString("subscriberId", statsItem.subscriberId);
            jsDataItem.putDouble("timestamp", statsItem.timeStamp);
            jsDataItem.putDouble("audioPacketsLost", statsItem.audioPacketsLost);
            jsDataItem.putDouble("audioBytesSent", statsItem.audioBytesSent);
            jsDataItem.putDouble("audioPacketsSent", statsItem.audioPacketsSent);
            jsStatsList.pushMap(jsDataItem);
        }
        
        eventData.putArray("data", jsStatsList);
        return eventData;
    }

    public static WritableMap prepareSubscriberVideoNetworkStats(SubscriberKit.SubscriberVideoStats stats) {

        WritableMap videoStats = Arguments.createMap();
        videoStats.putInt("videoPacketsLost", stats.videoPacketsLost);
        videoStats.putInt("videoBytesReceived", stats.videoBytesReceived);
        videoStats.putInt("videoPacketsReceived", stats.videoPacketsReceived);
        return videoStats;
    }

    public static WritableMap preparePublisherVideoNetworkStats(PublisherKit.PublisherVideoStats[] stats) {

        WritableMap eventData = Arguments.createMap();
        WritableArray jsStatsList = Arguments.createArray();

        for (PublisherKit.PublisherVideoStats statsItem : stats) {
            WritableMap jsDataItem = Arguments.createMap();
            jsDataItem.putString("connectionId", statsItem.connectionId);
            jsDataItem.putString("subscriberId", statsItem.subscriberId);
            jsDataItem.putDouble("timestamp", statsItem.timeStamp);
            jsDataItem.putDouble("videoPacketsLost", statsItem.videoPacketsLost);
            jsDataItem.putDouble("videoPacketsSent", statsItem.videoPacketsSent);
            jsDataItem.putDouble("videoBytesSent", statsItem.videoBytesSent);
            jsStatsList.pushMap(jsDataItem);
        }
        
        eventData.putArray("data", jsStatsList);
        return eventData;
    }

    public static WritableMap preparePublisherRTCStatsReport(PublisherKit.PublisherRtcStats[] rtcStatsReport) {

        WritableMap rtcStatsReportData = Arguments.createMap();
        WritableArray reports = Arguments.createArray();

        for (PublisherKit.PublisherRtcStats rtcStatsReportItem : rtcStatsReport) {
            WritableMap jsDataItem = Arguments.createMap();
            jsDataItem.putString("connectionId", rtcStatsReportItem.connectionId);
            jsDataItem.putString("jsonArrayOfReports", rtcStatsReportItem.jsonArrayOfReports);
            reports.pushMap(jsDataItem);
        }
        
        rtcStatsReportData.putArray("data", reports);
        return rtcStatsReportData;
    }

    public static WritableMap prepareSubscriberRTCStatsReport(String jsonArrayOfReports) {
        WritableMap rtcStatsReportData = Arguments.createMap();
        rtcStatsReportData.putString("data", jsonArrayOfReports);
        return rtcStatsReportData;
    }

    public static WritableMap createError(String message) {

        WritableMap errorInfo = Arguments.createMap();
        errorInfo.putString("message", message);
        return errorInfo;
    }
}

File: ios/OpenTokReactNative/Utils/EventUtils.swift
//
//  EventUtils.swift
//  OpenTokReactNative
//
//  Created by Manik Sachdeva on 11/3/18.
//  Copyright © 2018 TokBox Inc. All rights reserved.
//

import Foundation

class EventUtils {
    
    static var sessionPreface: String = "session:";
    static var publisherPreface: String = "publisher:";
    static var subscriberPreface: String = "subscriber:";
    
    static func prepareJSConnectionEventData(_ connection: OTConnection) -> Dictionary<String, Any> {
        var connectionInfo: Dictionary<String, Any> = [:];
        guard connection != nil else { return connectionInfo }
        connectionInfo["connectionId"] = connection.connectionId;
        connectionInfo["creationTime"] = convertDateToString(connection.creationTime);
        connectionInfo["data"] = connection.data;
        return connectionInfo;
    }
    
    static func prepareJSStreamEventData(_ stream: OTStream) -> Dictionary<String, Any> {
        var streamInfo: Dictionary<String, Any> = [:];
        guard OTRN.sharedState.sessions[stream.session.sessionId] != nil else { return streamInfo }
        streamInfo["streamId"] = stream.streamId;
        streamInfo["name"] = stream.name;
        streamInfo["connectionId"] = stream.connection.connectionId;
        streamInfo["connection"] = prepareJSConnectionEventData(stream.connection);
        streamInfo["hasAudio"] = stream.hasAudio;
        streamInfo["sessionId"] = stream.session.sessionId;
        streamInfo["hasVideo"] = stream.hasVideo;
        streamInfo["creationTime"] = convertDateToString(stream.creationTime);
        streamInfo["height"] = stream.videoDimensions.height;
        streamInfo["width"] = stream.videoDimensions.width;
        streamInfo["videoType"] = stream.videoType == OTStreamVideoType.screen ? "screen" : "camera"
        return streamInfo;
    }
    
    static func prepareJSErrorEventData(_ error: OTError) -> Dictionary<String, Any> {
        var errorInfo: Dictionary<String, Any> = [:];
        errorInfo["code"] = error.code;
        errorInfo["message"] = error.localizedDescription;
        return errorInfo;
    }
    
    static func prepareStreamPropertyChangedEventData(_ changedProperty: String, oldValue: Any, newValue: Any, stream: Dictionary<String, Any>) -> Dictionary<String, Any> {
        var streamPropertyEventData: Dictionary<String, Any> = [:];
        streamPropertyEventData["oldValue"] = oldValue;
        streamPropertyEventData["newValue"] = newValue;
        streamPropertyEventData["stream"] = stream;
        streamPropertyEventData["changedProperty"] = changedProperty;
        return streamPropertyEventData
    }
    
    static func prepareSubscriberVideoNetworkStatsEventData(_ videoStats: OTSubscriberKitVideoNetworkStats) -> Dictionary<String, Any> {
        var videoStatsEventData: Dictionary<String, Any> = [:];
        videoStatsEventData["videoPacketsLost"] = videoStats.videoPacketsLost;
        videoStatsEventData["videoBytesReceived"] = videoStats.videoBytesReceived;
        videoStatsEventData["videoPacketsReceived"] = videoStats.videoPacketsReceived;
        return videoStatsEventData;
    }
    
    static func preparePublisherVideoNetworkStatsEventData(_ videoStats: Array<OTPublisherKitVideoNetworkStats>) -> Dictionary<String, Any> {
        var eventData: Dictionary<String, Any> = [:];
        var stats = [Dictionary<String, Any>]()
        for statsItem in videoStats {
            var jsStatsItem: Dictionary<String, Any> = [:];
            jsStatsItem["connectionId"] = statsItem.connectionId;
            jsStatsItem["subscriberId"] = statsItem.subscriberId;
            jsStatsItem["timestamp"] = statsItem.timestamp;
            jsStatsItem["videoPacketsLost"] = statsItem.videoPacketsLost;
            jsStatsItem["videoPacketsSent"] = statsItem.videoPacketsSent;
            jsStatsItem["videoBytesSent"] = statsItem.videoBytesSent;
            stats.append(jsStatsItem)
        }
        eventData["data"] = stats
        return eventData;
    }
    
    static func prepareSubscriberAudioNetworkStatsEventData(_ audioStats: OTSubscriberKitAudioNetworkStats) -> Dictionary<String, Any> {
        var audioStatsEventData: Dictionary<String, Any> = [:];
        audioStatsEventData["audioPacketsLost"] = audioStats.audioPacketsLost;
        audioStatsEventData["audioBytesReceived"] = audioStats.audioBytesReceived;
        audioStatsEventData["audioPacketsReceived"] = audioStats.audioPacketsReceived;
        return audioStatsEventData;
    }
    
    static func prepareJSSessionEventData(_ session: OTSession) -> Dictionary<String, Any> {
        var sessionInfo: Dictionary<String, Any> = [:];
        sessionInfo["sessionId"] = session.sessionId;
        guard let connection = session.connection else { return sessionInfo };
        sessionInfo["connection"] = prepareJSConnectionEventData(connection);
        return sessionInfo;
    }
    
    static func prepareJSPublisherRTCStatsReport(_ rtcStatsReport: Array<OTPublisherRtcStats>) -> Dictionary<String, Any> {
        var rtcStatsReportData: Dictionary<String, Any> = [:];
        var reports = [Dictionary<String, Any>]()
        for rtcStatsReportItem in rtcStatsReport {
            var jsDataItem: Dictionary<String, Any> = [:];
            jsDataItem["connectionId"] = rtcStatsReportItem.connectionId
            jsDataItem["jsonArrayOfReports"] = rtcStatsReportItem.jsonArrayOfReports
            reports.append(jsDataItem)
        }
        rtcStatsReportData["data"] = reports
        return rtcStatsReportData;
    }
    
    static func prepareJSSubscriberRTCStatsReport(_ jsonArrayOfReports: String) -> Dictionary<String, Any> {
        var rtcStatsReportData: Dictionary<String, Any> = [:];
        rtcStatsReportData["data"] = jsonArrayOfReports
        return rtcStatsReportData;
    }
    
    static func getSupportedEvents() -> [String] {
        return [
            "\(sessionPreface)streamCreated",
            "\(sessionPreface)streamDestroyed",
            "\(sessionPreface)sessionDidConnect",
            "\(sessionPreface)sessionDidDisconnect",
            "\(sessionPreface)connectionCreated",
            "\(sessionPreface)connectionDestroyed",
            "\(sessionPreface)didFailWithError", 
            "\(sessionPreface)signal",
            "\(sessionPreface)archiveStartedWithId",
            "\(sessionPreface)archiveStoppedWithId",
            "\(sessionPreface)sessionDidBeginReconnecting",
            "\(sessionPreface)sessionDidReconnect",
            "\(sessionPreface)streamPropertyChanged",

            "\(publisherPreface)streamCreated",
            "\(publisherPreface)streamDestroyed",
            "\(publisherPreface)didFailWithError",
            "\(publisherPreface)audioLevelUpdated",
            "\(publisherPreface)videoNetworkStatsUpdated",
            "\(publisherPreface)rtcStatsReportUpdated",

            "\(subscriberPreface)subscriberDidConnect",
            "\(subscriberPreface)subscriberDidDisconnect",
            "\(subscriberPreface)didFailWithError",
            "\(subscriberPreface)videoNetworkStatsUpdated",
            "\(subscriberPreface)audioNetworkStatsUpdated",
            "\(subscriberPreface)audioLevelUpdated",
            "\(subscriberPreface)rtcStatsReportUpdated",
            "\(subscriberPreface)subscriberVideoEnabled",
            "\(subscriberPreface)subscriberVideoDisabled",
            "\(subscriberPreface)subscriberVideoDisableWarning",
            "\(subscriberPreface)subscriberVideoDisableWarningLifted",
            "\(subscriberPreface)subscriberVideoDataReceived",
            "\(subscriberPreface)subscriberDidReconnect"
        ];
    }
    
    static func convertDateToString(_ creationTime: Date) -> String {
        let dateFormatter: DateFormatter = DateFormatter();
        dateFormatter.dateFormat = "yyyy-MM-dd HH:mm:ss";
        dateFormatter.timeZone = TimeZone(abbreviation: "UTC");
        return dateFormatter.string(from:creationTime);
    }

    static func createErrorMessage(_ message: String) -> Dictionary<String, String> {
        var errorInfo: Dictionary<String, String> = [:]
        errorInfo["message"] = message
        return errorInfo
    }
    
}
